devito/equation.py

```diff
@@ -50,15 +56,32 @@ class Eq(sympy.Eq):
     def __new__(cls, *args, **kwargs):
         kwargs['evaluate'] = False
         subdomain = kwargs.pop('subdomain', None)
+        substitutions = kwargs.pop('coefficients', None)
         obj = sympy.Eq.__new__(cls, *args, **kwargs)
         obj._subdomain = subdomain
+        obj._substitutions = substitutions
+        functions = retrieve_functions(obj)
```

Comment:
If ``substitutions`` is None, then probably we don't need to spend time retrieving Functions. 
So we could:
```
if substituions is not None:
    functions = ...
    ...
```

This could save some time when big expressions are supplied (eg, when using high order derivatives)

Reply:
Ta, changed.

Reply:
Actually, since a function could be declared with symbolic coefficients but for a specific equation no replacement rules supplied I think this needs to stay.

Reply:
OK. Hopefully this should still return control the user in negligible time